scripts/jenkins/ardana/gerrit/gerrit_handle_event.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python
+
+import argparse
+import os
+import sys
+
+sys.path.append(os.path.dirname(__file__))
+from gerrit import GerritChange, GerritChangeSet  # noqa: E402
+from gerrit_merge import gerrit_merge  # noqa: E402
+from gerrit_review import gerrit_review  # noqa: E402
+
+
+def get_submittable_references(merged_change):
+    """
+    Get all submittable changes that directly depend on the supplied change
+    (have an explicit Depends-On referencing the indicated merged Gerrit
+    change and targeting the same branch, or that are submitted together
+    and directly on top of it).
+
+    :param merged_change: GerritChange object representing a merged change
+    :return: list of GerritChange object representing submittable changes,
+    referencing the supplied merged change as a direct dependency, and which
+    could be merged themselves
+    """
+
+    changeset = GerritChangeSet(
+        'is:open',
+        'is:mergeable',
+        'branch:{}'.format(merged_change.branch),
+        'label:Code-Review=2',
+        'label:Verified=2',
+        'label:Workflow=1',
+        '-label:Verified=-1',
+        '-label:Code-Review=-2'
+    )
+
+    references = []
+    for change in changeset.changes():
+        if change.submittable and change.depends_on_directly(merged_change):
+            references.append(change)
+
+    return references
+
+
+def get_stale_references(updated_change):
+    """
+    Get all changes that directly or indirectly (i.e. through a chain of
+    Depends-On commit message indicators) depend on the indicated updated
+    Gerrit change and targeting the same branch.
+
+    :param updated_change: GerritChange object representing a change
+    that has been updated (e.g. for which a new patchset was published)
+    :return: list of GerritChange object representing changes directly or
+    indirectly referencing the supplied updated change as a dependency,
+    which need to be re-verified
+    """
+    changeset = GerritChangeSet(
+        'is:open',
+        'branch:{}'.format(updated_change.branch),
+        'message:"Depends-On:"'
```

Comment:
This is very expensive as it requires loading in any change that has a "depends-on" in the commit regardless of its relevance. There isn't a lot to be done about this, but it does take a few queries per change, so it might be worth noting to look for optimisations. One of my earlier suggestions of using a Gerrit cache may help.

Reply:
This isn't as expensive as it might look, because no other queries need to be placed for the changes returned by this query, instead GerritChangeSet uses the returned result to create GerritChange objects without calling the Gerrit API again

Reply:
Doesn't it call a version of the get_related? Either way, loading the object in as you are is very good :-).

Reply:
I implemented a simple REST API caching, which will help greatly with that.